### DISCUSSION TITLE
Using a 3D velocity model via netcdf

### DISCUSSION THREAD

**Discussion by ArCSchmidt** [SeisSol/SeisSol#1104] [Q&A]
## Problem 
Reading a 3D velocity model  netcdf using ASAGI. The program is being terminated failing to find models for certain points. A .zip of the relevant files is attached ([netcdf_issue.zip](https://github.com/SeisSol/SeisSol/files/15237519/netcdf_issue.zip)), if necessary the full launch directory can be found here (https://nextcloud.ifg.uni-kiel.de/index.php/s/K46qLdrp2HzQfTi). 
 ## Description
I am a new SeisSol user and do not know how to proceed.
I am trying to set up a simple test for using a 3D velocity model in SeisSol. I have tried using the !ASAGI map (materials.yaml), as well as an affine map as described in the SeisSol documentation (materials_affine.yaml). Both approaches provide different error messages but ultimately terminate (v0.3.err and v0.3_affine.err respectively).

With !ASAGI the error looks like this (repeating):
```
Tue May 07 16:48:40, Warn:  ASAGI: Coordinate in dimension 0  is out of range. Fixing. 
Tue May 07 16:48:40, Warn:  ASAGI: Coordinate in dimension 2  is out of range. Fixing. 
Tue May 07 16:48:40, Warn:  ASAGI: Coordinate in dimension 0  is out of range. Fixing. 
Tue May 07 16:48:40, Warn:  ASAGI: Coordinate in dimension 2  is out of range. Fixing. 
Tue May 07 16:48:40, Warn:  ASAGI: Coordinate in dimension 0  is out of range. Fixing. 
```
With !AffineMap the error looks like this (repeating):
```
  what():  materials_affine.yaml@1: Could not find model for point [ 746923 4.51192e+06 -13339.9 ] in group 1.
[n119:2738560] *** Process received signal ***
[n119:2738560] Signal: Aborted (6)
[n119:2738560] Signal code:  (-6)
[n119:2738560] [ 0] /lib64/libpthread.so.0(+0x12ce0)[0x14bfaaa0ace0]
[n119:2738560] [ 1] /lib64/libc.so.6(gsignal+0x10f)[0x14bfaa681a4f]
[n119:2738560] [ 2] /lib64/libc.so.6(abort+0x127)[0x14bfaa654db5]
[n119:2738560] [ 3] /work_beegfs/sungw886/fresh_seissol/spack-0.20.0/opt/spack/linux-rocky8-x86_64/gcc-8.5.0/gcc-12.3.0-amrnqjjktnk62z3xwr6437cxt2en2lw2/lib64/libstdc++.so.6(+0xa9ff9)[0x14bfab262ff9]
[n119:2738560] [ 4] /work_beegfs/sungw886/fresh_seissol/spack-0.20.0/opt/spack/linux-rocky8-x86_64/gcc-8.5.0/gcc-12.3.0-amrnqjjktnk62z3xwr6437cxt2en2lw2/lib64/libstdc++.so.6(+0xb561a)[0x14bfab26e61a]
[n119:2738560] [ 5] /work_beegfs/sungw886/fresh_seissol/spack-0.20.0/opt/spack/linux-rocky8-x86_64/gcc-8.5.0/gcc-12.3.0-amrnqjjktnk62z3xwr6437cxt2en2lw2/lib64/libstdc++.so.6(+0xb5685)[0x14bfab26e685]
[n119:2738560] [ 6] /work_beegfs/sungw886/fresh_seissol/spack-0.20.0/opt/spack/linux-rocky8-x86_64/gcc-8.5.0/gcc-12.3.0-amrnqjjktnk62z3xwr6437cxt2en2lw2/lib64/libstdc++.so.6(+0xb58d7)[0x14bfab26e8d7]
[n119:2738560] [ 7] ./SeisSol_Release_dhsw_4_elastic[0x41edc3]
[n119:2738560] [ 8] ./SeisSol_Release_dhsw_4_elastic[0x5f8ddc]
[n119:2738560] [ 9] ./SeisSol_Release_dhsw_4_elastic[0x50f1d9]
[n119:2738560] [10] ./SeisSol_Release_dhsw_4_elastic[0x5414a1]
[n119:2738560] [11] ./SeisSol_Release_dhsw_4_elastic[0x59d932]
[n119:2738560] [12] ./SeisSol_Release_dhsw_4_elastic[0x5a131e]
[n119:2738560] [13] ./SeisSol_Release_dhsw_4_elastic[0x58c149]
[n119:2738560] [14] ./SeisSol_Release_dhsw_4_elastic[0x558f5c]
[n119:2738560] [15] ./SeisSol_Release_dhsw_4_elastic[0x424a9e]
[n119:2738560] [16] ./SeisSol_Release_dhsw_4_elastic[0x4222d3]
[n119:2738560] [17] /lib64/libc.so.6(__libc_start_main+0xf3)[0x14bfaa66dca3]
[n119:2738560] [18] ./SeisSol_Release_dhsw_4_elastic[0x4246ae]
[n119:2738560] *** End of error message ***
```
Seemingly both approaches fail to find a model for certain points even though the extent of the netcdf should encompass all points in the mesh. The test uses an adaptation of TPV5 from the cookbook.

It would be great if a more experienced user could tell me what i am doing wrong and point me in the right direction.

**Comment by hsudha** [comment]
I am working with ArCSchmidt. After a personally communicated suggestion that the order of values in the netcdf could play a role - z order should be ascending - we tried all variations of sorting. Unfortunately, the described behavior and result remained. After a while Seissol stops because it is missing a coordinate that to our knowledge is within the mesh. What to do? 
We are very grateful for hints and suggestions! Thanks!

**Comment by Thomas-Ulrich** [comment]
Hi,
my guess is that ASAGI only supports (and assumes) the same dx for x,y and z.
(therefore it might compute dx from x[1]-x[0] and use it for y and z).
And by the way, your coordinates are large values, and you may have problems (not related with current issue) when running with floats because of https://github.com/SeisSol/SeisSol/issues/200

**Reply by hsudha** [reply]
Ah! Thomas, thanks a lot for these hints! We will try to adapt the inputs to cubic voxels, if I understand you correctly. The issue with the coordinates, well, we are using UTM meters, but we could go to km. We will inform this channel about the outcome.

**Reply by Thomas-Ulrich** [reply]
you can just translate everything by a constant.

**Reply by hsudha** [reply]
Hi Thomas and everyone,
we did try the proposed ends to sort out the problem. We tried UTM coordinates in kilometers, rather than meter, changed the reference point for smaller coordinate values. We re-gridded the model to have cubic voxels. However, still we don't get the model to work. Seissol starts working on the problem and stops after 5 minutes or so with a complaint about coordinates.

We had the idea to compare what we want to do with the Seissol examples that are based on community 3D velocity models. Here, unfortunately, the download options of these models are broken. 

Would it be possible to provide a launch folder with a  working example containing a correct netcdf velocity model for us to try out with our installation? In this way we hopefully could do the forensics about the differences in files on our side. This would be most helpful at this point, I believe. Or is there a better way? We appreciate any help. Thanks you!

**Reply by Thomas-Ulrich** [reply]
Hi, you can have a look at :
https://github.com/SeisSol/Training/blob/main/kaikoura/NZ_rhomulambda_large.yaml
https://github.com/SeisSol/Training/tree/main/kaikoura/asagi

**Reply by hsudha** [reply]
Hi, Thomas, thanks a lot for pointing us to the training example! The only difference we spot is that your coordinates are in `float32`, ours in `float64`. This is easy to change and to test. We will inform you about the result.
Cheers, Henriette
